ImageInLib/ImageInLib/src/imageInterpolation.c

```diff
@@ -3,257 +3,145 @@
+    return sqrt(pow(point1.x - point2.x, 2) + pow(point1.y - point2.y, 2));
+}
 
-    size_t i, j, k, in, jn, kn, indx;
+Point2D getNearestNeighbor2D(Point2D point, PixelSpacing spacing) {
 
-    size_t lengthNew = length * 2, widthNew = width * 2, heightNew = height * 2;
+    PointNeighbors2D neighbors;
 
-    for (k = 0; k < height; k++) {
-        for (i = 0; i < length; i++) {
-            for (j = 0; j < width; j++) {
+    neighbors = getPointNeighbors2D(point, spacing);
 
-                indx = x_new(i, j, length);
-                in = 2 * i; jn = j * 2; kn = k * 2;
+    dataType distance[4];
+    distance[0] = getDistance2D(point, neighbors.top_left);
+    distance[1] = getDistance2D(point, neighbors.top_right);
+    distance[2] = getDistance2D(point, neighbors.bottom_left);
+    distance[3] = getDistance2D(point, neighbors.bottom_right);
 
-                newImage[kn][x_new(in, jn, lengthNew)] = originalImage[k][indx];
-                newImage[kn][x_new(in + 1, jn, lengthNew)] = originalImage[k][indx];
-                newImage[kn][x_new(in, jn + 1, lengthNew)] = originalImage[k][indx];
-                newImage[kn][x_new(in + 1, jn + 1, lengthNew)] = originalImage[k][indx];
+    dataType min_dist = 1000000000000;
+    int k, min_indice;
 
-                newImage[kn + 1][x_new(in, jn, lengthNew)] = originalImage[k][indx];
-                newImage[kn + 1][x_new(in + 1, jn, lengthNew)] = originalImage[k][indx];
-                newImage[kn + 1][x_new(in, jn + 1, lengthNew)] = originalImage[k][indx];
-                newImage[kn + 1][x_new(in + 1, jn + 1, lengthNew)] = originalImage[k][indx];
-            }
+    for (k = 0; k < 4; k++) {
+        if (distance[k] < min_dist) {
+            min_dist = distance[k];
+            min_indice = k;
         }
     }
 
-    return true;
-}
-
-/*
- * Get 3D point in real world from 3D image point
- * Three operations are done here : image scaling, image translation and image rotation
-*/
-Point3D imageCoordToRealCoord(Point3D srcPoint, Point3D imageOrigin, VoxelSpacing imageSpacing, OrientationMatrix orientation) {
-    Point3D resultPoint;
-    resultPoint.x = imageOrigin.x + imageSpacing.sx * (orientation.v1.x + orientation.v1.y + orientation.v1.z) * srcPoint.x;
-    resultPoint.y = imageOrigin.y + imageSpacing.sy * (orientation.v2.x + orientation.v2.y + orientation.v2.z) * srcPoint.y;
-    resultPoint.z = imageOrigin.z + imageSpacing.sz * (orientation.v3.x + orientation.v3.y + orientation.v3.z) * srcPoint.z;
-    return resultPoint;
+    if (min_indice == 0) {
+        return neighbors.top_left;
+    }
+    if (min_indice == 1) {
+        return neighbors.top_right;
+    }
+    if (min_indice == 2) {
+        return neighbors.bottom_left;
+    }
+    if (min_indice == 3) {
+        return neighbors.bottom_right;
+    }
 }
 
+bool interpolateImageFromImageCStoRealWorldCS(Image_Data2D src_image, Image_Data2D interp_image, dataType scale_factor) {
```

Comment:
I still do not understand, why we need scaling factor...?
In my opinion, we do not.

Reply:
I need this scaling factor because, for me, when we are doing interpolation, the source image and the destination image don't have the same number of pixels, that is exactly why we are interpolating. The pixel spacing is used when we need to compute distance between points/pixels in order to find the closest one for example.

Reply:
AFAIK scale_factor is used commonly in cases of visualisation, when one is scaling the image (just transforming image or its part to arbitrary  size - with potentially new origin and dimension ).
Let us consider particular case (let us consider same orientations).
We have:
source image:
origin: 100, 100, pxSpacing: 2,2, dimension: 100,100
then image_src(0,0) -> rw_src(100,100)
image_src(1,1) -> rw_src(102, 102) ...
image_src(99,99) -> rw_src(298.298)
interpolated image { origin 10,10, pxSpacing: 1,1, dimensions: 200,200}
then image_int(0,0) -> rw_int(10,10)
image_int(1,1) -> rw_int(11, 11) ...
image_int(199,199) -> rw_int(209,209)
So we exactly know, what are the dimesions of particular  images.

When we need to find image_src corresponding to image_int we can do it without scale_factor

e.g image_int.x = 199 -> rw_int = 10 + 199 * 1= 209 -> img_src = (209 - 100) / (2 * 1) = 109 / 2 = 54.5
same for image_int_y...

so the corresponding point to image_int(199,199) is  image_src(54.5, 54.5)
We have not needed any scale factor for it. Even if the src and int dimensions are different.

Reply:
So, if I understood correctly, we do have to different images with different origin, spacing and dimension (and probably different orientation) both in real world coordinates system. Then we want to interpolate, let say, the first image to have same dimension as the second image using all this information ? I am right?